File: days/src/main/java/aoc2024j/day5/Day5.java

package aoc2024j.day5;

import utils.IO;

import java.util.*;
import java.util.function.Predicate;

class Rules {
    Map<Integer, Set<Integer>> afters = new HashMap<>();

    void parseAndAdd(String line) {
        var parts = line.split("\\|");
        var before = Integer.parseInt(parts[0]);
        var after = Integer.parseInt(parts[1]);
        this.afters.computeIfAbsent(before, k -> new HashSet<>()).add(after);
    }

    long part1(Updates update) {
        return update.updates.stream()
                .filter(this::countValid)
                .mapToLong(Update::middle)
                .sum();
    }

    private boolean countValid(Update update) {
        for (int i = 0; i < update.pages.size() - 1; i++) {
            var validAfters = this.afters.get(update.pages.get(i));
            if (validAfters == null) {
                return false;
            }
            for (int j = i + 1; j < update.pages.size(); j++) {
                if (!validAfters.contains(update.pages.get(j))) {
                    return false;
                }
            }
        }
        return true;
    }

    Comparator<Integer> comparator() {
        return (i, j) -> {
            if (i.equals(j)) {
                return 0;
            }
            var aftersOfI = this.afters.getOrDefault(i, Set.of());
            if (aftersOfI.contains(j)) {
                return -1;
            }
            return 1;
        };
    }

    long part2(Updates update) {
        return update.updates.stream()
                .filter(Predicate.not(this::countValid))
                .map(u -> Update.sort(u, this.comparator()))
                .mapToLong(Update::middle)
                .sum();
    }
}

class Update {
    final List<Integer> pages;

    Update(List<Integer> pages) {
        if (pages.size() % 2 != 1) {
            throw new IllegalArgumentException("Expected odd pages, got " + pages.size());
        }
        this.pages = pages;
    }

    static Update parse(String line) {
        var parts = line.split(",");
        var list = Arrays.stream(parts)
                .map(Integer::parseInt)
                .toList();
        return new Update(list);
    }

    public long middle() {
        return pages.get(pages.size() / 2);
    }

    static Update sort(Update update, Comparator<Integer> comparator) {
        var sorted = new ArrayList<>(update.pages);
        sorted.sort(comparator);
        return new Update(sorted);
    }
}

class Updates {
    List<Update> updates = new ArrayList<>();

    void parseAndAdd(String line) {
        this.updates.add(Update.parse(line));
    }

}

record ParseResult(Rules rules, Updates updates) {
    static ParseResult parse(List<String> data) {
        var rules = new Rules();
        var updates = new Updates();
        for (var line : data) {
            if (line.isBlank()) {
                continue;
            }
            if (line.contains("|")) {
                rules.parseAndAdd(line);
            } else {
                updates.parseAndAdd(line);
            }
        }
        return new ParseResult(rules, updates);
    }
}

public class Day5 {

    long part1(List<String> data) {
        var parsed = ParseResult.parse(data);
        return parsed.rules().part1(parsed.updates());
    }

    long part2(List<String> data) {
        var parsed = ParseResult.parse(data);
        return parsed.rules().part2(parsed.updates());
    }

    public static void main(String[] args) {
        var day5 = new Day5();
        var data = IO.getResourceAsList("aoc2024/day5.txt");
        var part1 = day5.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day5.part2(data);
        System.out.println("part2 = " + part2);
    }
}

File: days/src/main/scala/aoc2024s/day5/Day5.scala

package aoc2024s.day5;

import utils.IO

import scala.jdk.CollectionConverters.*
import scala.util.chaining.scalaUtilChainingOps

object Day5 {

  case class Rules(afters: Map[Int, Set[Int]]) {

    def part1(updates: Updates): Long =
      updates.updates
        .filter(isValid)
        .map(_.middle)
        .sum

    def isValid(update: Update): Boolean =
      update.pages
        .tails
        .filter(_.nonEmpty)
        .forall: suffix =>
          suffix.tail
            .forall: page =>
              afters.getOrElse(suffix.head, Set.empty).contains(page)

    def ordering: Ordering[Int] = (i, j) =>
      if (i == j) 0
      else if (afters.getOrElse(i, Set.empty).contains(j)) -1
      else 1


    def part2(updates: Updates): Long =
      updates.updates
        .filterNot(isValid)
        .map(_.sorted(ordering))
        .map(_.middle)
        .sum
  }

  object Rules {

    def parse(ruleLines: List[String]): Rules =
      ruleLines
        .map(line => line.split('|').map(_.toInt))
        .groupMapReduce(_(0))(a => Set(a(1)))(_ ++ _)
        .pipe(Rules.apply)
  }

  case class Updates(updates: List[Update])

  object Updates {

    def parse(updateLines: List[String]): Updates =
      updateLines
        .map(Update.parse)
        .pipe(Updates.apply)
  }

  case class Update(pages: Vector[Int]) {
    def middle: Long = pages(pages.size / 2).toLong

    def sorted(ordering: Ordering[Int]): Update =
      Update(pages.sorted(ordering))
  }

  object Update {
    def parse(line: String): Update =
      Update(line.split(',').map(_.toInt).toVector)
  }

  def parse(data: List[String]): (Rules, Updates) = {
    val (ruleLines, updateLines) = data.span(_.nonEmpty);
    val rules = Rules.parse(ruleLines);
    val updates = Updates.parse(updateLines.tail);
    (rules, updates);
  }

  def part1(data: List[String]): Long = {
    val (rules, updates) = parse(data);
    rules.part1(updates);
  }

  def part2(data: List[String]): Long = {
    val (rules, updates) = parse(data);
    rules.part2(updates);
  }

  @main def main5(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day5.txt").asScala.toList;
    val part1 = Day5.part1(data);
    println(s"part1 = $part1");
    val part2 = Day5.part2(data);
    println(s"part2 = $part2");
  }
}